### ISSUE TITLE
Pyright not finding editable package in monorepo context

### ISSUE THREAD

**Issue by cheog** [microsoft/pyright#4116]
**Describe the bug**
**Tl;dr** Pyright not picking up on an editable package that's outside the immediate project.

I've already read about how import hooks mess this up but I checked pyrights documentation and it says pdm does this the old, compatible way by default and I double checked pdm's documentation and it also says that it does it the old compatible way.

Im using pdm as the package manager and have a monorepo that looks as follows
```
.
├── sub_project1
│   ├── project1
│   │   └── __init__.py
│   └── pyproject.toml
└── sub_project2
    ├── __pypackages__
    ├── pdm.lock
    ├── project2
    │   └── __init__.py
    └── pyproject.toml
```
```
# project1/__init__.py
def func1(a: int, b: int):
    return a + b
```
```
# project2/__init__.py
from project1 import func
```

I'm getting `Import "project1" could not be resolved` in `project2/__init__.py`

**Expected behavior**
I expect it to recognize projec1

**To Reproduce**
Here is the full project tree including `__pypackages__`
```
.
├── sub_project1
│   ├── project1
│   │   └── __init__.py
│   └── pyproject.toml
└── sub_project2
    ├── __pypackages__
    │   └── 3.9
    │       ├── bin
    │       ├── include
    │       └── lib
    │           ├── project1-0.1.0.dist-info
    │           │   ├── METADATA
    │           │   ├── RECORD
    │           │   ├── WHEEL
    │           │   └── direct_url.json
    │           ├── project1.pth
    │           ├── project2-0.2.0.dist-info
    │           │   ├── METADATA
    │           │   ├── RECORD
    │           │   ├── WHEEL
    │           │   └── direct_url.json
    │           └── project2.pth
    ├── pdm.lock
    ├── project2
    │   └── __init__.py
    └── pyproject.toml
```
and the output from `pyright --verbose project2/__init__.py` ran from `sub_project2`
```
No configuration file found.
pyproject.toml file found at /Users/me/scratch/monorepo/sub_project2.
Loading pyproject.toml file at /Users/me/scratch/monorepo/sub_project2/pyproject.toml
Assuming Python version 3.9
Assuming Python platform Darwin
Auto-excluding **/node_modules
Auto-excluding **/__pycame__
Auto-excluding **/.*
stubPath /Users/me/scratch/monorepo/sub_project2/typings is not a valid directory.
Search paths for /Users/me/scratch/monorepo/sub_project2
  /Users/me/.npm/_npx/f1c7b3657b214e2f/node_modules/pyright/dist/typeshed-fallback/stdlib
  /Users/me/scratch/monorepo/sub_project2
  /Users/me/scratch/monorepo/sub_project2/__pypackages__/3.9/lib
  /Users/me/scratch/monorepo/sub_project2/typings
  /Users/me/.npm/_npx/f1c7b3657b214e2f/node_modules/pyright/dist/typeshed-fallback/stubs/...
  /Users/me/Library/Python/3.10/lib/python/site-packages/pdm/pep582
  /opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9
  /opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload
  /opt/homebrew/lib/python3.9/site-packages
Searching for source files
Found 1 source file
Could not import 'project1' in file '/Users/me/scratch/monorepo/sub_project2/project2/__init__.py'
  Looking in stubPath '/Users/me/scratch/monorepo/sub_project2/typings'
  Attempting to resolve stub package using root path '/Users/me/scratch/monorepo/sub_project2/typings'
  Attempting to resolve using root path '/Users/me/scratch/monorepo/sub_project2/typings'
  Looking in root directory of execution environment '/Users/me/scratch/monorepo/sub_project2'
  Attempting to resolve stub package using root path '/Users/me/scratch/monorepo/sub_project2'
  Attempting to resolve using root path '/Users/me/scratch/monorepo/sub_project2'
  Looking in extraPath '/Users/me/scratch/monorepo/sub_project2/__pypackages__/3.9/lib'
  Attempting to resolve stub package using root path '/Users/me/scratch/monorepo/sub_project2/__pypackages__/3.9/lib'
  Attempting to resolve using root path '/Users/me/scratch/monorepo/sub_project2/__pypackages__/3.9/lib'
  Looking in python search path '/Users/me/Library/Python/3.10/lib/python/site-packages/pdm/pep582'
  Attempting to resolve stub package using root path '/Users/me/Library/Python/3.10/lib/python/site-packages/pdm/pep582'
  Attempting to resolve using root path '/Users/me/Library/Python/3.10/lib/python/site-packages/pdm/pep582'
  Looking in python search path '/opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9'
  Attempting to resolve stub package using root path '/opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9'
  Attempting to resolve using root path '/opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9'
  Looking in python search path '/opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload'
  Attempting to resolve stub package using root path '/opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload'
  Attempting to resolve using root path '/opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload'
  Looking in python search path '/opt/homebrew/lib/python3.9/site-packages'
  Attempting to resolve stub package using root path '/opt/homebrew/lib/python3.9/site-packages'
  Attempting to resolve using root path '/opt/homebrew/lib/python3.9/site-packages'
  Looking for typeshed stdlib path
  Looking for typeshed stdlib path
  Attempting to resolve using root path '/Users/me/.npm/_npx/f1c7b3657b214e2f/node_modules/pyright/dist/typeshed-fallback/stdlib'
  Typeshed path not found
  Looking for typeshed third-party path
  Looking for typeshed stubs path
  Typeshed path not found
pyright 1.1.277
/Users/me/scratch/monorepo/sub_project2/project2/__init__.py
  /Users/me/scratch/monorepo/sub_project2/project2/__init__.py:1:6 - error: Import "project1" could not be resolved (reportMissingImports)
```
And the contents of the pyproject.tomls
```
[project]
name = "project2"
version = "0.2.0"
description = ""
authors = [
    {name = "Me", email = "me@gmail.com"},
]
dependencies = [
]
requires-python = ">=3.9"
license = {text = "MIT"}
[project.optional-dependencies]

[build-system]
requires = ["pdm-pep517>=1.0.0"]
build-backend = "pdm.pep517.api"

[tool.pdm.dev-dependencies]
dev = [
    "-e file:///${PROJECT_ROOT}/../sub_project1#egg=project1",
]

[tool.pyright]
extraPaths = ["__pypackages__/3.9/lib"]
```
```
[project]
name = "project1"
version = "0.1.0"
description = ""
authors = [
    {name = "Me", email = "me@gmail.com"},
]
dependencies = []
requires-python = ">=3.9"
license = {text = "MIT"}

[build-system]
requires = ["pdm-pep517>=1.0.0"]
build-backend = "pdm.pep517.api"

```

**VS Code extension or command-line**
I'm using neovim and coc-pyright and I reinstalled it this morning.

**Comment by erictraut** [comment]
Please refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md) for details about how pyright resolves imports.

Based on what I see in the log, pyright is working as expected here.

From your logs above, I can see that `sub_project1` is not included in your search paths (which are dumped from `sys.path` in the configured Python environment). That means project1 is not installed using a compatible editable install mechanism. Prior to the recent unfortunate changes related to PEP 660, editable installs used ".pth" files that augmented `sys.path` to point to the installed library.

You've configured `extraPaths` to point to an absolute path that's outside of your project, which is odd. `extraPaths` is designed to be used with relative paths located under your project root. Have you tried simply adding "sub_project1" to your `extraPaths`?

**Comment by cheog** [comment]
I have tried adding "sub_project1" to the `extraPaths` and while that alleviates the immediate issue it causes problems because both projects have `tests` folders. Also, more generally they should be independent so I don't want to be worrying about naming clashes and such (especially because we're about to add quite a few more libraries to this monorepo).

It is using the ".pth" files. The contents of `project1.pth` is `/Users/me/scratch/monorepo/sub_project1` by the way. [Your own docs](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md#pdm) also say that it should be installed in the correct way.

`extraPaths` is pointing to a relative path that is inside the project directory (it is set to `"__pypackages__/3.9/lib"`, it's not absolute as it doesn't start with a / and it's inside the project directory when cwd is `sub_project2` which is what it always is). I use the `extraPaths` because without it pyright doesn't pick up on the whole `__pypackages__` thing. I always have to do my editing from `sub_project2` specifically because the extraPaths entry is relative (if I'm in `sub_project2/project2` for example, pyright no longer picks up on `__pypackages__`)

**Comment by erictraut** [comment]
Ah yes, I was wrong about `extraPaths` pointing to an absolute path. OK, so that's fine.

The problem here is that `sub_project1` is not part of `sys.path` in your Python environment. I would expect it to be included in `sys.path` if there's a ".pth" present and it contains `/Users/me/scratch/monorepo/sub_project1`. You'll need to debug that and determine why this path isn't in `sys.path`. By default, pyright determines the search paths by simply invoking the selected Python interpreter and dumping the value of `sys.path`.

**Comment by erictraut** [comment]
I'm going to close this issue for now because I don't think it's a bug in pyright. If your debugging attempts reveal some additional clues that seem to indicate a pyright bug, please let me know and I'll reopen.
